File: algorithm/39. Combination Sum/39.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> result;
        vector<int> path;
        sort(candidates.begin(),candidates.end());
        helper(candidates,0,target,path,result);
        return result;
    }
    void helper(vector<int> &nums,int index,int target,vector<int>& path,vector<vector<int>> & result)
    {
        if(target==0)
        {
            result.push_back(path);
            return;
        }
        for(int i=index;i<nums.size();i++)
        {
            if(nums[i]>target)
                break;
            path.push_back(nums[i]);
            helper(nums,i,target-nums[i],path,result);
            path.pop_back();
        }
    }
};

File: algorithm/39. Combination Sum/39.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        Arrays.sort(candidates);
        helper(candidates, target, res, new ArrayList<Integer>(), 0);
        return res;
    }
    public void helper(int[] candidates, int target, List<List<Integer>> res, List<Integer> com, int index)
    {
        if(target == 0)
        {
            res.add(new ArrayList<Integer>(com));
            return;
        }
        for(int i = index; i < candidates.length && target >= candidates[i]; i++)
        {
            com.add(candidates[i]);
            helper(candidates, target - candidates[i], res, com,i);
            com.remove(com.size() - 1);
        }
        return;
    }
}
